src/sys_task_manager.c

```diff
@@ -0,0 +1,236 @@
+
+static void SysTaskManager_InitTask(SysTaskManager *sysTaskMgr, SysTask *task);
+static void SysTaskManager_InitTasks(SysTaskManager *sysTaskMgr);
+static SysTask *SysTaskManager_AllocTask(SysTaskManager *sysTaskMgr);
+static BOOL SysTaskManager_FreeTask(SysTaskManager *sysTaskMgr, SysTask *task);
+static SysTask *SysTaskManager_InternalAddTask(SysTaskManager *sysTaskMgr, SysTaskFunc callback, void *param, u32 priority);
+
+static void SysTaskManager_InitTask(SysTaskManager *sysTaskMgr, SysTask *task)
+{
+    task->manager = sysTaskMgr;
+    task->prevTask = task->nextTask = &sysTaskMgr->sentinelTask;
+    task->priority = 0;
+    task->param = NULL;
+    task->callback = NULL;
+}
+
+static void SysTaskManager_InitTasks(SysTaskManager *sysTaskMgr)
+{
+    for (int i = 0; i < sysTaskMgr->maxTasks; i++) {
+        SysTaskManager_InitTask(sysTaskMgr, &sysTaskMgr->tasks[i]);
+        sysTaskMgr->taskStack[i] = &sysTaskMgr->tasks[i];
+    }
+
+    sysTaskMgr->stackPointer = 0;
+}
+
+static SysTask *SysTaskManager_AllocTask(SysTaskManager *sysTaskMgr)
+{
+    if (sysTaskMgr->stackPointer == sysTaskMgr->maxTasks) {
+        return NULL;
+    }
+
+    SysTask *task = sysTaskMgr->taskStack[sysTaskMgr->stackPointer];
+    sysTaskMgr->stackPointer++;
+
+    return task;
+}
+
+static BOOL SysTaskManager_FreeTask(SysTaskManager *sysTaskMgr, SysTask *task)
+{
+    if (sysTaskMgr->stackPointer == 0) {
+        return FALSE;
+    }
+
+    task->manager = sysTaskMgr;
+    task->prevTask = task->nextTask = &sysTaskMgr->sentinelTask;
+    task->priority = 0;
+    task->param = NULL;
+    task->callback = NULL;
+
+    sysTaskMgr->stackPointer--;
+    sysTaskMgr->taskStack[sysTaskMgr->stackPointer] = task;
+
+    return TRUE;
+}
+
+u32 SysTaskManager_GetRequiredSize(u32 maxTasks)
+{
+    return sizeof(SysTaskManager) + (sizeof(SysTask *) + sizeof(SysTask)) * maxTasks;
+}
+
+SysTaskManager *SysTaskManager_Init(u32 maxTasks, void *memory)
+{
+    GF_ASSERT(memory);
+    SysTaskManager *sysTaskMgr = memory;
+
+    sysTaskMgr->taskStack = (SysTask **)((u8 *)(sysTaskMgr) + sizeof(SysTaskManager));
+    sysTaskMgr->tasks = (SysTask *)((u8 *)(sysTaskMgr->taskStack) + sizeof(SysTask *) * maxTasks);
+    sysTaskMgr->maxTasks = maxTasks;
+    sysTaskMgr->stackPointer = 0;
+    sysTaskMgr->isBeingModified = FALSE;
+
+    SysTaskManager_InternalInit(sysTaskMgr);
+    return sysTaskMgr;
+}
+
+void SysTaskManager_InternalInit(SysTaskManager *sysTaskMgr)
+{
+    SysTaskManager_InitTasks(sysTaskMgr);
+
+    sysTaskMgr->sentinelTask.manager = sysTaskMgr;
+    sysTaskMgr->sentinelTask.prevTask = sysTaskMgr->sentinelTask.nextTask = &sysTaskMgr->sentinelTask;
+    sysTaskMgr->sentinelTask.priority = 0;
+    sysTaskMgr->sentinelTask.param = NULL;
+    sysTaskMgr->sentinelTask.callback = NULL;
+
+    sysTaskMgr->currentTask = sysTaskMgr->sentinelTask.nextTask;
+}
+
+void SysTaskManager_ExecuteTasks(SysTaskManager *sysTaskMgr)
+{
+    if (sysTaskMgr->isBeingModified) {
```

Comment:
**suggestion:** based on this behavior, maybe this member should be named `locked`